boards: st: stm32f4_disco: add audio output

titouanc opened the pull request:
The stm32f4_disco board features an audio codec (cs43l22) connected to an audio jack (TRS) output, and to the stm32f4 via I2S3 (sound) and I2C1 (control).

First, add a new audio codec driver for the cs43l22. This initial driver implementation only has the bare minimum initialization for the stm32f4_disco. In particular, many control registers are not initialized by the driver, as their default values just work on this board. Also, I couldn't test the speaker output, because it is not wired here (only headphones).

Moreover, to set this up, several peripherals must be configured:

- Enable I2S3 and PLLI2S (tuned for 22.05/44.1/88.2 kHz audio)
- Enable DMA1 (used by I2S3)
- Enable I2C1, and add definition for the CS43L22 on the bus
- Disable USART1 and CAN1 which have conflicting pinctrl with I2C1

This new pin setting is more appropriate, because this gives access to the audio jack output by default (instead of "bare" pins not connected to anything on the board).

Finally, add support for the stm32f4_disco in the sample `i2s_codec` to demonstrate usage of the newly introduced audio output feature on this board

titouanc committed:
audio: codec: add new cs43l22 driver

Add new driver for the Cirrus CS43L22 audio DAC.

Signed-off-by: Titouan Christophe <moiandme@gmail.com>

titouanc committed:
boards: st: stm32f4_disco: add support for audio output

The stm32f4_disco board features an audio codec (cs43l22) connected to
an audio jack (TRS) output, and to the stm32f4 via I2S and I2C. To set
this up, several peripherals must be configured:

- Enable I2S3 and the PLLI2S (tuned for 22.05/44.1/88.2 kHz audio)
- Enable DMA1 (used by I2S3)
- Enable I2C1, and add definition for the CS43L22 on the bus
- Disable USART1 and CAN1 which have conflicting pinctrl with I2C1

This new setting is more appropriate, because this gives access to the
audio jack output by default (instead of "bare" pins not connected to
anything on the board).

Signed-off-by: Titouan Christophe <moiandme@gmail.com>

erwango commented on boards/st/stm32f4_disco/stm32f4_disco.dts:
```diff
@@ -187,3 +174,39 @@ zephyr_udc0: &usbotg_fs {
+	 * - 44.1kHz  / 16b, 24b or 32b
+	 * - 88.2kHz  / 16b or 24b
+	 */
+	mul-n = <271>;
```
It may be useful to also specifies that it takes same input as main PLL and hence this one should not be changed w/o checking impact on PLLI2S

erwango reviewed (commented):
Looks perfectly fine, and changing default config makes sense. This being said, it could be worth specifying in migration guide.
Also, did you check that the configurations you're removing are not used in tests or samples ? In which case, .overlay files should be added/updated.

Last, is there any sample in tree that would help to build the new driver ? At least `build_all` would be required.

titouanc commented:
> Last, is there any sample in tree that would help to build the new driver ?

Yes indeed, the sample [i2s_codec](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/drivers/i2s/i2s_codec) is a good example. However, this sample allocates a lot of sound buffers, which don't fit into the stm32f407 RAM, so we have to configure a much lower audio sample rate. This ultimately defeats the 44.1kHz optimal tuning introduced on PLLI2S. But I will add support for the stm32f4_disco to this sample as a 3rd commit to this PR so at least, we have a working example in tree.

> At least build_all would be required.

I don't know what `build_all` is in this context. Would you have an example, a link to the docs or any pointer for me to check ?

erwango commented:
> I don't know what `build_all` is in this context. Would you have an example, a link to the docs or any pointer for me to check ?

These are specific tests aiming at at east compiling drivers that don't have dedicated samples. See `tests/drivers/build_all`

titouanc committed:
samples: drivers: i2s: i2s_codec: add support for stm32f4_disco

Add configuration and board overlay in the I2S audio codec sample to
demonstrate usage of the newly introduced audio output feature.

Signed-off-by: Titouan Christophe <moiandme@gmail.com>

titouanc commented:
Thank you for your review @erwango !

I just added support for the stm32f4_disco in the sample `i2s_codec` (and checked that it works here). Let me know if you want me to add it in `tests/drivers/build_all` as well.

I also added an additional note in the dt overlay to check for changes in the main PLL, and rebased on top of the latest `main`

titouanc commented:
And as a side note, I've checked that no other sample is actually using CAN1 or USART1. The following command doesn't output anything:

```bash
for f in $(find . -name stm32f4_disco.overlay); do
    grep can1 $f
    grep usart1 $f
done
```

gautierg-st reviewed (changes requested):
One small comment about clock. Otherwise LGTM

titouanc commented on boards/st/stm32f4_disco/stm32f4_disco.dts:
```diff
@@ -187,3 +174,41 @@ zephyr_udc0: &usbotg_fs {
+	mck-enabled;
+	pinctrl-0 = <&i2s3_ws_pa4 &i2s3_ck_pc10 &i2s3_sd_pc12 &i2s3_mck_pc7>;
+	pinctrl-names = "default";
+	clocks = <&rcc STM32_CLOCK(APB1, 15U)>, <&rcc STM32_SRC_PLLI2S_R NO_SEL>;
```
Thanks ! Change applied and rebased on top of main
